File: puppeteer-python/fullteaching_e2e/common/browser_user.py
# -*- coding: utf-8 -*-
"""A browser + page pair for one simulated end user, ported from common/BrowserUser.java
and common/ChromeUser.java.

Puppeteer (and therefore pyppeteer) only automates Chromium, so unlike the Java suite
there is no ChromeUser/FirefoxUser/EdgeUser split - BrowserUser always launches Chromium.
"""
import logging
import os

from pyppeteer import launch

logger = logging.getLogger(__name__)

CHROME = "chrome"
FIREFOX = "firefox"
EDGE = "edge"

_CHROMIUM_ARGS = [
    "--start-maximized",
    "--use-fake-ui-for-media-stream",
    "--use-fake-device-for-media-stream",
    "--ignore-certificate-errors",
]

# Injected once per page so tests can probe "is a <video> actually playing" via JS,
# mirrors the GLOBAL_JS_FUNCTION literal duplicated in BaseLoggedTest.setup()/setupBrowser().
_VIDEO_PLAYING_PROBE_JS = (
    "window.MY_FUNC = function(containerQuerySelector) {"
    "var elem = document.createElement('div');"
    "elem.id = 'video-playing-div';"
    "elem.innerText = 'VIDEO PLAYING';"
    "document.body.appendChild(elem);"
    "console.log('Video check function successfully added to DOM by Puppeteer')};"
)


class BrowserUser:
    """Wraps one Chromium browser + page, mirroring common/BrowserUser.java's role."""

    def __init__(self, client_data: str, timeout_seconds: int, test_name: str):
        self.client_data = client_data
        self.timeout_ms = timeout_seconds * 1000
        self.test_name = test_name
        self.is_on_session = False
        self.browser = None
        self.page = None
        self.app_url = None

    async def _configure(self, headless: bool):
        logger.info("Starting the configuration of the Chromium browser for %s", self.client_data)
        self.browser = await launch(
            headless=headless,
            args=_CHROMIUM_ARGS,
            ignoreHTTPSErrors=True,
            executablePath=os.environ.get("PYPPETEER_EXECUTABLE_PATH"),
        )
        self.page = await self.browser.newPage()
        logger.info("Driver successfully configured for %s", self.client_data)

    async def run_javascript(self, script: str, *args):
        return await self.page.evaluate(script, *args)

    async def wait_until(self, awaitable_factory, error_message: str):
        """Awaits `awaitable_factory()`, mirrors BrowserUser.waitUntil(condition, errorMessage)."""
        try:
            return await awaitable_factory()
        except TimeoutError as timeout:
            logger.error(error_message)
            raise TimeoutError(f'"{error_message}" (checked with condition) > {timeout}') from timeout

    async def dispose(self):
        if self.browser is not None:
            await self.browser.close()


async def _resolve_browser_key(browser: str) -> str:
    if browser not in (CHROME, FIREFOX, EDGE):
        logger.warning("Unknown browser key '%s', defaulting to Chromium", browser)
    elif browser != CHROME:
        logger.warning("pyppeteer only automates Chromium; ignoring requested browser '%s'", browser)
    return CHROME


async def setup_browser(app_url: str, test_name: str, user_identifier: str, seconds_of_wait: int,
                         browser: str = CHROME, headless: bool = True) -> BrowserUser:
    """Creates a BrowserUser, navigates it to app_url and injects the video-playing probe.

    Mirrors BaseLoggedTest.setupBrowser(browser, testName, userIdentifier, secondsOfWait).
    """
    await _resolve_browser_key(browser)
    user = BrowserUser(user_identifier, seconds_of_wait, test_name)
    await user._configure(headless=headless)
    user.app_url = app_url
    logger.info("Navigating to %s", app_url)
    await user.page.goto(app_url)
    await user.run_javascript(_VIDEO_PLAYING_PROBE_JS)
    return user

File: playwright-csharp/Common/BrowserUser.cs
using Microsoft.Playwright;

namespace FullTeaching.E2E.Common;

/// <summary>
/// One simulated end user: an isolated Playwright <see cref="IBrowserContext"/> + <see cref="IPage"/>,
/// ported from common/BrowserUser.java and common/ChromeUser.java.
///
/// Unlike Selenium (one WebDriver process per user) and pyppeteer (one Chromium process per
/// user), Playwright's browser contexts are cheap, fully isolated (separate cookies/storage/
/// permissions) sessions within a *single* shared browser process - the documented Playwright
/// pattern for exactly this "simulate several independent logged-in users" scenario, which is
/// why (unlike the Cypress port) EChatTests/VideoSessionTests/LoggedVideoSessionTests are
/// fully implemented here rather than skipped: Playwright can drive several simultaneous
/// sessions where Cypress fundamentally cannot.
/// </summary>
public class BrowserUser
{
    private const string VideoPlayingProbeJs = """
        window.MY_FUNC = function(containerQuerySelector) {
            var elem = document.createElement('div');
            elem.id = 'video-playing-div';
            elem.innerText = 'VIDEO PLAYING';
            document.body.appendChild(elem);
            console.log('Video check function successfully added to DOM by Playwright');
        };
        """;

    public string ClientData { get; }
    public int TimeoutMs { get; }
    public string TestName { get; }
    public bool IsOnSession { get; set; }
    public string AppUrl { get; private set; } = string.Empty;
    public IBrowserContext Context { get; private set; } = null!;
    public IPage Page { get; private set; } = null!;

    private BrowserUser(string clientData, int timeoutSeconds, string testName)
    {
        ClientData = clientData;
        TimeoutMs = timeoutSeconds * 1000;
        TestName = testName;
    }

    /// <summary>
    /// Creates a BrowserUser, navigates it to appUrl and injects the video-playing probe.
    /// Mirrors BaseLoggedTest.setupBrowser(browser, testName, userIdentifier, secondsOfWait).
    /// </summary>
    public static async Task<BrowserUser> SetupBrowserAsync(
        IBrowser browser, string appUrl, string testName, string userIdentifier, int secondsOfWait)
    {
        var user = new BrowserUser(userIdentifier, secondsOfWait, testName)
        {
            AppUrl = appUrl,
        };

        user.Context = await browser.NewContextAsync(new BrowserNewContextOptions
        {
            IgnoreHTTPSErrors = true,
            Permissions = ["camera", "microphone"],
        });
        user.Context.SetDefaultTimeout(user.TimeoutMs);
        user.Page = await user.Context.NewPageAsync();

        await user.Page.GotoAsync(appUrl);
        await user.Page.EvaluateAsync(VideoPlayingProbeJs);
        return user;
    }

    public Task<object?> RunJavascriptAsync(string script, object? arg = null) =>
        Page.EvaluateAsync<object?>(script, arg);

    /// <summary>
    /// Awaits <paramref name="action"/>, mirrors BrowserUser.waitUntil(condition, errorMessage):
    /// re-throws Playwright's own timeout with the caller's more specific error message prefixed.
    /// </summary>
    public async Task WaitUntilAsync(Func<Task> action, string errorMessage)
    {
        try
        {
            await action();
        }
        catch (PlaywrightException ex)
        {
            throw new PlaywrightException($"\"{errorMessage}\" (checked with condition) > {ex.Message}", ex);
        }
    }

    public async Task<T> WaitUntilAsync<T>(Func<Task<T>> action, string errorMessage)
    {
        try
        {
            return await action();
        }
        catch (PlaywrightException ex)
        {
            throw new PlaywrightException($"\"{errorMessage}\" (checked with condition) > {ex.Message}", ex);
        }
    }

    public async Task DisposeAsync()
    {
        await Context.CloseAsync();
    }
}
